scilpy/reconst/multi_processes.py

```diff
@@ -0,0 +1,420 @@
+        is the largest peak.
+        Default: 0.5
+    absolute_threshold : float, optional
+        Absolute threshold on fODF amplitude. This value should be set to
+        approximately 1.5 to 2 times the maximum fODF amplitude in isotropic
+        voxels (ex. ventricles). The script `compute_fodf_max_in_ventricles.py`
+        can be used to find the maximal value.
+        Default: 0
+    min_separation_angle : float in [0, 90], optional
+        The minimum distance between directions. If two peaks are too close
+        only the larger of the two is returned.
+        Default: 25
+    normalize_peaks : bool, optional
+        If true, all peak values are calculated relative to `max(odf)`.
+    npeaks : int, optional
+        Maximum number of peaks found (default 5 peaks).
+    sh_basis_type : str, optional
+        Type of spherical harmonic basis used for `shm_coeff`. Either
+        `descoteaux07` or `tournier07`.
+        Default: `descoteaux07`
+    nbr_processes: int, optional
+        The number of subprocesses to use.
+        Default: multiprocessing.cpu_count()
+
+    Returns
+    -------
+    tuple of np.ndarray
+        peak_dirs, peak_values, peak_indices
+    """
+    sh_order = order_from_ncoef(shm_coeff.shape[-1])
+    B, _ = sh_to_sf_matrix(sphere, sh_order, sh_basis_type)
+
+    data_shape = shm_coeff.shape
+    if mask is not None:
+        shm_coeff = applymask(shm_coeff, mask)
+
+    nbr_processes = multiprocessing.cpu_count() if nbr_processes is None \
+        or nbr_processes < 0 else nbr_processes
+
+    shm_coeff = shm_coeff.ravel().reshape(np.prod(data_shape[0:3]), data_shape[3])
+    chunks = np.array_split(shm_coeff, nbr_processes)
+    chunk_len = np.cumsum([0] + [len(c) for c in chunks])
+
+    pool = multiprocessing.Pool(nbr_processes)
+    results = pool.map(peaks_from_sh_parallel,
+                       zip(chunks,
+                           itertools.repeat(B),
+                           itertools.repeat(sphere),
+                           itertools.repeat(relative_peak_threshold),
+                           itertools.repeat(absolute_threshold),
+                           itertools.repeat(min_separation_angle),
+                           itertools.repeat(npeaks),
+                           itertools.repeat(normalize_peaks),
+                           np.arange(len(chunks))))
+    pool.close()
+    pool.join()
+
+    peak_dirs_array = np.zeros((np.prod(data_shape[0:3]), npeaks, 3))
+    peak_values_array = np.zeros((np.prod(data_shape[0:3]), npeaks))
+    peak_indices_array = np.zeros((np.prod(data_shape[0:3]), npeaks))
+    for i, peak_dirs, peak_values, peak_indices in results:
+        peak_dirs_array[chunk_len[i]:chunk_len[i+1], :] = peak_dirs
+        peak_values_array[chunk_len[i]:chunk_len[i+1], :] = peak_values
+        peak_indices_array[chunk_len[i]:chunk_len[i+1], :] = peak_indices
+
+    peak_dirs_array = peak_dirs_array.reshape(data_shape[0:3]+(npeaks, 3))
```

Comment:
A comment explaining this is to reshape the outputs so they are back to their original spatial position.